src/context.rs

```diff
@@ -10,6 +10,211 @@ use crate::ffi::types::{c_uint, c_void, AlignedType};
 use crate::ffi::{self, CPtr};
 use crate::{Error, Secp256k1};
 
+#[cfg(not(feature = "std"))]
+mod internal {
+    use core::cell::UnsafeCell;
+    use core::hint::spin_loop;
+    use core::marker::PhantomData;
+    use core::mem::ManuallyDrop;
+    use core::ops::{Deref, DerefMut};
+    use core::ptr::NonNull;
+    use core::sync::atomic::{AtomicBool, Ordering};
+
+    use crate::ffi::types::{c_void, AlignedType};
+    use crate::{ffi, AllPreallocated, Context, Secp256k1};
+
+    const MAX_SPINLOCK_ATTEMPTS: usize = 128;
+    const MAX_PREALLOC_SIZE: usize = 16; // measured at 208 bytes on Andrew's 64-bit system
+
+    static SECP256K1: SpinLock = SpinLock::new();
+
+    // Simple spinlock-gated structure which holds the backing store for a
+    // secp256k1 context.
+    //
+    // To obtain exclusive access, call [`Self::try_lock`], which will spinlock
+    // for some small number of iterations before giving up. By trying again in
+    // a loop, you can emulate a "true" spinlock that will only yield once it
+    // has access. However, this would be very dangerous, especially in a nostd
+    // environment, because if we are pre-empted by an interrupt handler while
+    // the lock is held, and that interrupt handler attempts to take the lock,
+    // then we deadlock.
+    //
+    // Instead, the strategy we take within this module is to simply create a
+    // new stack-local context object if we are unable to obtain a lock on the
+    // global one. This is slow and loses the defense-in-depth "rerandomization"
+    // anti-sidechannel measure, but it is better than deadlocking..
+    struct SpinLock {
+        flag: AtomicBool,
+        // Invariant: if this is non-None, then the store is valid and can be
+        // used with `ffi::secp256k1_context_preallocated_create`.
+        data: UnsafeCell<([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>)>,
+    }
+
+    // Required by rustc if we have a static of this type.
+    // Safety: `data` is accessed only while the `flag` is held.
+    unsafe impl Sync for SpinLock {}
+    unsafe impl Send for SpinLock {}
+
+    impl SpinLock {
+        const fn new() -> Self {
+            Self {
+                flag: AtomicBool::new(false),
+                data: UnsafeCell::new(([AlignedType::ZERO; MAX_PREALLOC_SIZE], None)),
```

Comment:
Maybe this should be `MaybeUninit` instead? (Yeah, it's tiny perf difference but better semantic clarity, I think.)